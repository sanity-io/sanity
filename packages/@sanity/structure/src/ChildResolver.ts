import {CollectionBuilder, Collection, SerializePath, SerializeOptions} from './StructureNodes'

export interface ChildResolverOptions {
  index: number
  parent: Collection
  parentPath?: SerializePath
  parameters?: {[key: string]: any}
  serializeOptions?: SerializeOptions
}

export type ItemChild = CollectionBuilder | Collection | Function | undefined

interface ChildObservable {
  subscribe: (child: ItemChild | Promise<ItemChild>) => {}
}

export interface ChildResolver {
  (itemId: string, options: ChildResolverOptions):
    | ItemChild
    | Promise<ItemChild>
    | ChildObservable
    | undefined
}
