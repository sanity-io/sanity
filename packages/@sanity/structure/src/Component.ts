import {camelCase} from 'lodash'
import {SerializeOptions, StructureNode, Serializable, Child} from './StructureNodes'
import {SerializeError, HELP_URL} from './SerializeError'
import {MenuItem, MenuItemBuilder, maybeSerializeMenuItem} from './MenuItem'
import {MenuItemGroup, MenuItemGroupBuilder, maybeSerializeMenuItemGroup} from './MenuItemGroup'
import {validateId} from './util/validateId'
import {FixMe} from './types'

export interface Component extends StructureNode {
  component: FixMe
  child?: Child
  menuItems: MenuItem[]
  menuItemGroups: MenuItemGroup[]
  options: {[key: string]: unknown}
}

export interface ComponentInput extends StructureNode {
  component: FixMe
  child?: Child
  options?: {[key: string]: unknown}
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export interface BuildableComponent extends Partial<StructureNode> {
  component?: FixMe
  child?: Child
  options?: {[key: string]: unknown}
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export class ComponentBuilder implements Serializable {
  protected spec: BuildableComponent

  constructor(spec?: ComponentInput) {
    this.spec = {options: {}, ...(spec ? spec : {})}
  }

  id(id: string) {
    return this.clone({id})
  }

  getId() {
    return this.spec.id
  }

  title(title: string) {
    return this.clone({title, id: this.spec.id || camelCase(title)})
  }

  getTitle() {
    return this.spec.title
  }

  child(child: Child) {
    return this.clone({child})
  }

  getChild() {
    return this.spec.child
  }

  component(component: FixMe) {
    return this.clone({component})
  }

  getComponent() {
    return this.spec.component
  }

  options(options: {[key: string]: unknown}) {
    return this.clone({options})
  }

  getOptions() {
    return this.spec.options || {}
  }

  menuItems(menuItems: (MenuItem | MenuItemBuilder)[]) {
    return this.clone({menuItems})
  }

  getMenuItems() {
    return this.spec.menuItems
  }

  menuItemGroups(menuItemGroups: (MenuItemGroup | MenuItemGroupBuilder)[]) {
    return this.clone({menuItemGroups})
  }

  getMenuItemGroups() {
    return this.spec.menuItemGroups
  }

  serialize(options: SerializeOptions = {path: []}): Component {
    const {id, title, child, options: componentOptions, component} = this.spec
    if (!id) {
      throw new SerializeError(
        '`id` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!component) {
      throw new SerializeError(
        '`component` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    return {
      id: validateId(id, options.path, options.index),
      title,
      type: 'component',
      child,
      component,
      options: componentOptions || {},
      menuItems: (this.spec.menuItems || []).map((item, i) =>
        maybeSerializeMenuItem(item, i, options.path)
      ),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) =>
        maybeSerializeMenuItemGroup(item, i, options.path)
      ),
    }
  }

  clone(withSpec?: BuildableComponent): ComponentBuilder {
    const builder = new ComponentBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}
    return builder
  }
}
