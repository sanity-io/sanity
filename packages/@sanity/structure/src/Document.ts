import {uniq, camelCase} from 'lodash'
import {SerializeOptions, Serializable, Child, DocumentNode, EditorNode} from './StructureNodes'
import {SerializeError, HELP_URL} from './SerializeError'
import {SchemaType} from './parts/Schema'
import {validateId} from './util/validateId'
import {View, ViewBuilder, maybeSerializeView} from './views/View'
import {form} from './views'
import {
  getUserDefinedDefaultDocumentBuilder,
  DocumentFragmentResolveOptions
} from './userDefinedStructure'
import {getTemplateById} from '@sanity/initial-value-templates'

interface DocumentOptions {
  id: string
  type: string
  template?: string
  templateParameters?: {
    [key: string]: any
  }
}

export type PartialDocumentNode = {
  id?: string
  title?: string
  child?: Child
  views?: (View | ViewBuilder)[]
  options?: Partial<DocumentOptions>
}

export class DocumentBuilder implements Serializable {
  protected spec: PartialDocumentNode

  constructor(spec?: PartialDocumentNode) {
    this.spec = spec ? spec : {}
  }

  id(id: string): DocumentBuilder {
    return this.clone({id})
  }

  getId() {
    return this.spec.id
  }

  title(title: string) {
    return this.clone({title, id: this.spec.id || camelCase(title)})
  }

  getTitle() {
    return this.spec.title
  }

  child(child: Child) {
    return this.clone({child})
  }

  getChild() {
    return this.spec.child
  }

  documentId(documentId: string): DocumentBuilder {
    // Let's try to be a bit helpful and assign an ID from document ID if none is specified
    const paneId = this.spec.id || documentId
    return this.clone({
      id: paneId,
      options: {
        ...(this.spec.options || {}),
        id: documentId
      }
    })
  }

  getDocumentId() {
    return this.spec.options && this.spec.options.id
  }

  schemaType(documentType: SchemaType | string): DocumentBuilder {
    return this.clone({
      options: {
        ...(this.spec.options || {}),
        type: typeof documentType === 'string' ? documentType : documentType.name
      }
    })
  }

  getSchemaType() {
    return this.spec.options && this.spec.options.type
  }

  initialValueTemplate(templateId: string, parameters?: {[key: string]: any}) {
    return this.clone({
      options: {
        ...(this.spec.options || {}),
        template: templateId,
        templateParameters: parameters
      }
    })
  }

  getInitalValueTemplate() {
    return this.spec.options && this.spec.options.template
  }

  getInitialValueTemplateParameters() {
    return this.spec.options && this.spec.options.templateParameters
  }

  views(views: (View | ViewBuilder)[]) {
    return this.clone({views})
  }

  getViews(): (View | ViewBuilder)[] {
    return this.spec.views || []
  }

  serialize({path = [], index, hint}: SerializeOptions = {path: []}): DocumentNode {
    const urlId = path[index || path.length - 1]

    // Try to grab document ID / editor ID from URL if not defined
    const id = this.spec.id || (urlId && `${urlId}`) || ''
    const options: Partial<DocumentOptions> = {
      id,
      type: undefined,
      template: undefined,
      templateParameters: undefined,
      ...this.spec.options
    }

    if (typeof id !== 'string' || !id) {
      throw new SerializeError(
        '`id` is required for document nodes',
        path,
        index,
        hint
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!options || !options.id) {
      throw new SerializeError(
        'document id (`id`) is required for document nodes',
        path,
        id,
        hint
      ).withHelpUrl(HELP_URL.DOCUMENT_ID_REQUIRED)
    }

    const views = (this.spec.views && this.spec.views.length > 0 ? this.spec.views : [form()]).map(
      (item, i) => maybeSerializeView(item, i, path)
    )

    const viewIds = views.map(view => view.id)
    const dupes = uniq(viewIds.filter((id, i) => viewIds.includes(id, i + 1)))
    if (dupes.length > 0) {
      throw new SerializeError(
        `document node has views with duplicate IDs: ${dupes.join(',  ')}`,
        path,
        id,
        hint
      )
    }

    return {
      ...this.spec,
      child: this.spec.child,
      id: validateId(id, path, index),
      type: 'document',
      options: getDocumentOptions(options),
      views
    }
  }

  clone(withSpec: PartialDocumentNode = {}) {
    const builder = new DocumentBuilder()
    const options = {...(this.spec.options || {}), ...(withSpec.options || {})}
    builder.spec = {...this.spec, ...withSpec, options}
    return builder
  }
}

function getDocumentOptions(spec: Partial<DocumentOptions>): DocumentOptions {
  const opts: DocumentOptions = {
    id: spec.id || '',
    type: spec.type || '*'
  }

  if (spec.template) {
    opts.template = spec.template
  }

  if (spec.templateParameters) {
    opts.templateParameters = spec.templateParameters
  }

  return opts
}

export function documentFromEditor(spec?: EditorNode) {
  let doc =
    spec && spec.type
      ? // Use user-defined document fragment as base if possible
        getDefaultDocumentNode({schemaType: spec.type})
      : // Fall back to plain old document builder
        new DocumentBuilder()

  if (spec) {
    const {id, type, template, templateParameters} = spec.options

    doc = doc.id(spec.id).documentId(id)

    if (type) {
      doc = doc.schemaType(type)
    }

    if (template) {
      doc = doc.initialValueTemplate(template, templateParameters)
    }

    if (spec.child) {
      doc = doc.child(spec.child)
    }
  }

  return doc
}

export function documentFromEditorWithInitialValue(
  templateId: string,
  parameters?: {[key: string]: any}
) {
  const template = getTemplateById(templateId)
  if (!template) {
    throw new Error(`Template with ID "${templateId}" not defined`)
  }

  return getDefaultDocumentNode({schemaType: template.schemaType}).initialValueTemplate(
    templateId,
    parameters
  )
}

export function getDefaultDocumentNode(
  options: DocumentFragmentResolveOptions
): DocumentBuilder {
  const {documentId, schemaType} = options
  const userDefined = getUserDefinedDefaultDocumentBuilder(options)

  let builder = userDefined || new DocumentBuilder()
  if (!builder.getId()) {
    builder = builder.id('documentEditor')
  }

  if (documentId) {
    builder = builder.documentId(documentId.replace(/^drafts\./, ''))
  }

  return builder.schemaType(schemaType)
}
