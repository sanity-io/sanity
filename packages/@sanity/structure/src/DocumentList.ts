import {
  getParameterlessTemplatesBySchemaType,
  getTemplateById,
} from '@sanity/initial-value-templates'
import {SchemaType, getDefaultSchema} from './parts/Schema'
import {isActionEnabled} from './parts/documentActionUtils'
import {resolveTypeForDocument} from './util/resolveTypeForDocument'
import {SortItem} from './Sort'
import {SerializeError, HELP_URL} from './SerializeError'
import {SerializeOptions, Child} from './StructureNodes'
import {ChildResolver, ChildResolverOptions, ItemChild} from './ChildResolver'
import {InitialValueTemplateItem} from './InitialValueTemplateItem'
import {
  GenericListBuilder,
  BuildableGenericList,
  GenericList,
  GenericListInput,
} from './GenericList'
import {DocumentBuilder, getDefaultDocumentNode} from './Document'

const validateFilter = (spec: PartialDocumentList, options: SerializeOptions) => {
  const filter = spec.options!.filter.trim()

  if (['*', '{'].includes(filter[0])) {
    throw new SerializeError(
      `\`filter\` cannot start with \`${filter[0]}\` - looks like you are providing a query, not a filter`,
      options.path,
      spec.id,
      spec.title
    ).withHelpUrl(HELP_URL.QUERY_PROVIDED_FOR_FILTER)
  }

  return filter
}

const resolveDocumentChildForItem: ChildResolver = (
  itemId: string,
  options: ChildResolverOptions
): ItemChild | Promise<ItemChild> | undefined => {
  const parentItem = options.parent as DocumentList
  const template = options.params?.template ? getTemplateById(options.params.template) : undefined
  const type = template
    ? template.schemaType
    : parentItem.schemaTypeName || resolveTypeForDocument(itemId)

  return Promise.resolve(type).then((schemaType) =>
    schemaType
      ? getDefaultDocumentNode({schemaType, documentId: itemId})
      : new DocumentBuilder().id('editor').documentId(itemId).schemaType('')
  )
}

export interface PartialDocumentList extends BuildableGenericList {
  options?: DocumentListOptions
  schemaTypeName?: string
}

export interface DocumentListInput extends GenericListInput {
  options: DocumentListOptions
}

export interface DocumentList extends GenericList {
  options: DocumentListOptions
  child: Child
  schemaTypeName?: string
}

interface DocumentListOptions {
  filter: string
  params?: {[key: string]: any}
  apiVersion?: string
  defaultOrdering?: SortItem[]
}

export class DocumentListBuilder extends GenericListBuilder<
  PartialDocumentList,
  DocumentListBuilder
> {
  protected spec: PartialDocumentList

  constructor(spec?: DocumentListInput) {
    super()
    this.spec = spec ? spec : {}
    this.initialValueTemplatesSpecified = Boolean(spec && spec.initialValueTemplates)
  }

  apiVersion(apiVersion: string): DocumentListBuilder {
    return this.clone({options: {...(this.spec.options || {filter: ''}), apiVersion}})
  }

  getApiVersion(): string | undefined {
    return this.spec.options?.apiVersion
  }

  filter(filter: string): DocumentListBuilder {
    return this.clone({options: {...(this.spec.options || {}), filter}})
  }

  getFilter(): string | undefined {
    return this.spec.options?.filter
  }

  schemaType(type: SchemaType | string): DocumentListBuilder {
    const schemaTypeName = typeof type === 'string' ? type : type.name
    return this.clone({schemaTypeName})
  }

  getSchemaType(): string | undefined {
    return this.spec.schemaTypeName
  }

  params(params: Record<string, unknown>): DocumentListBuilder {
    return this.clone({
      options: {...(this.spec.options || {filter: ''}), params},
    })
  }

  getParams(): Record<string, unknown> | undefined {
    return this.spec.options?.params
  }

  defaultOrdering(ordering: SortItem[]): DocumentListBuilder {
    if (!Array.isArray(ordering)) {
      throw new Error('`defaultOrdering` must be an array of order clauses')
    }

    return this.clone({
      options: {...(this.spec.options || {filter: ''}), defaultOrdering: ordering},
    })
  }

  getDefaultOrdering(): SortItem[] | undefined {
    return this.spec.options?.defaultOrdering
  }

  serialize(options: SerializeOptions = {path: []}): DocumentList {
    if (typeof this.spec.id !== 'string' || !this.spec.id) {
      throw new SerializeError(
        '`id` is required for document lists',
        options.path,
        options.index,
        this.spec.title
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!this.spec.options || !this.spec.options.filter) {
      throw new SerializeError(
        '`filter` is required for document lists',
        options.path,
        this.spec.id,
        this.spec.title
      ).withHelpUrl(HELP_URL.FILTER_REQUIRED)
    }

    return {
      ...super.serialize(options),
      type: 'documentList',
      schemaTypeName: this.spec.schemaTypeName,
      child: this.spec.child || resolveDocumentChildForItem,
      options: {
        ...this.spec.options,
        apiVersion:
          this.spec.options.apiVersion ||
          // If this is a simple type filter, use modern API version - otherwise default to v1
          (this.spec.options?.filter === '_type == $type' ? '2021-06-07' : '1'),
        filter: validateFilter(this.spec, options),
      },
    }
  }

  clone(withSpec?: PartialDocumentList): DocumentListBuilder {
    const builder = new DocumentListBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}

    // Since initialValueTemplatesSpecified is only set in the constructor, we need to
    // manually copy it over to not override initialValueTemplates below
    if (this.initialValueTemplatesSpecified) {
      builder.initialValueTemplatesSpecified = true
    }

    if (!this.initialValueTemplatesSpecified) {
      builder.spec.initialValueTemplates = inferInitialValueTemplates(builder.spec)
    }

    if (!builder.spec.schemaTypeName) {
      builder.spec.schemaTypeName = inferTypeName(builder.spec)
    }

    return builder
  }

  getSpec(): PartialDocumentList {
    return this.spec
  }
}

function inferInitialValueTemplates(
  spec: PartialDocumentList
): InitialValueTemplateItem[] | undefined {
  const schema = getDefaultSchema()
  const {schemaTypeName, options} = spec
  const {filter, params} = options || {filter: '', params: {}}
  const typeNames = schemaTypeName ? [schemaTypeName] : getTypeNamesFromFilter(filter, params)

  if (typeNames.length === 0) {
    return undefined
  }

  const templateItems: InitialValueTemplateItem[] = []
  return typeNames.reduce((items, typeName) => {
    const schemaType = schema.get(typeName)
    if (!isActionEnabled(schemaType, 'create')) {
      return items
    }

    return items.concat(
      getParameterlessTemplatesBySchemaType(typeName).map(
        (tpl): InitialValueTemplateItem => ({
          type: 'initialValueTemplateItem',
          id: tpl.id,
          templateId: tpl.id,
        })
      )
    )
  }, templateItems)
}

function inferTypeName(spec: PartialDocumentList): string | undefined {
  const {options} = spec
  const {filter, params} = options || {filter: '', params: {}}
  const typeNames = getTypeNamesFromFilter(filter, params)
  return typeNames.length === 1 ? typeNames[0] : undefined
}

export function getTypeNamesFromFilter(
  filter: string,
  params: {[key: string]: any} = {}
): string[] {
  let typeNames = getTypeNamesFromEqualityFilter(filter, params)

  if (typeNames.length === 0) {
    typeNames = getTypeNamesFromInTypesFilter(filter, params)
  }

  return typeNames
}

// From _type == "movie" || _type == $otherType
function getTypeNamesFromEqualityFilter(
  filter: string,
  params: {[key: string]: any} = {}
): string[] {
  const pattern = /\b_type\s*==\s*(['"].*?['"]|\$.*?(?:\s|$))|\B(['"].*?['"]|\$.*?(?:\s|$))\s*==\s*_type/g
  const matches: string[] = []
  let match
  while ((match = pattern.exec(filter)) !== null) {
    matches.push(match[1] || match[2])
  }

  return matches
    .map((candidate) => {
      const typeName = candidate[0] === '$' ? params[candidate.slice(1)] : candidate
      const normalized = (typeName || '').trim().replace(/^["']|["']$/g, '')
      return normalized
    })
    .filter(Boolean)
}

// From _type in ["dog", "cat", $otherSpecies]
function getTypeNamesFromInTypesFilter(
  filter: string,
  params: {[key: string]: any} = {}
): string[] {
  const pattern = /\b_type\s+in\s+\[(.*?)\]/
  const matches = filter.match(pattern)
  if (!matches) {
    return []
  }

  return matches[1]
    .split(/,\s*/)
    .map((match) => match.trim().replace(/^["']+|["']+$/g, ''))
    .map((item) => (item[0] === '$' ? params[item.slice(1)] : item))
    .filter(Boolean)
}
