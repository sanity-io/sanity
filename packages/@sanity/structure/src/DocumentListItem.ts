import {SerializeOptions} from './StructureNodes'
import {Partial} from './Partial'
import {ListItemBuilder, ListItem, UnserializedListItem, ListItemInput} from './ListItem'
import {SchemaType} from './parts/Schema'
import {SerializeError, HELP_URL} from './SerializeError'
import {DocumentBuilder, getDefaultDocumentNode} from './Document'

export interface DocumentListItemInput extends ListItemInput {
  schemaType: SchemaType | string
}

export interface DocumentListItem extends ListItem {
  schemaType: SchemaType
  _id: string
}

type PartialDocumentListItem = Partial<UnserializedListItem>

const getDefaultChildResolver = (spec: PartialDocumentListItem) => (documentId: string) => {
  const schemaType =
    spec.schemaType &&
    (typeof spec.schemaType === 'string' ? spec.schemaType : spec.schemaType.name)

  return schemaType
    ? getDefaultDocumentNode({schemaType, documentId})
    : new DocumentBuilder().id('documentEditor').documentId(documentId)
}

export class DocumentListItemBuilder extends ListItemBuilder {
  protected spec: PartialDocumentListItem

  constructor(spec?: DocumentListItemInput) {
    super(spec)
    this.spec = spec ? spec : {}
  }

  serialize(options: SerializeOptions = {path: []}): DocumentListItem {
    const spec = super.serialize({...options, titleIsOptional: true})

    if (!spec.schemaType) {
      throw new SerializeError(
        '`schemaType` is required for document list items',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.SCHEMA_TYPE_REQUIRED)
    }

    const child = spec.child || getDefaultChildResolver(spec)
    return {...spec, child, schemaType: spec.schemaType, _id: spec.id}
  }

  clone(withSpec?: PartialDocumentListItem) {
    const builder = new DocumentListItemBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}
    return builder
  }
}
