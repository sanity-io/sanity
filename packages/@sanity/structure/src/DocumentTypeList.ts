import {DocumentListBuilder, DocumentListInput, PartialDocumentList} from './DocumentList'
import {Child} from './StructureNodes'
import {DEFAULT_INTENT_HANDLER} from './Intent'
import {GenericListInput} from './GenericList'
import {SchemaType} from './parts/Schema'

export interface DocumentTypeListInput extends Partial<GenericListInput> {
  schemaType: SchemaType | string
}

export class DocumentTypeListBuilder extends DocumentListBuilder {
  protected spec: PartialDocumentList

  constructor(spec?: DocumentListInput) {
    super()
    this.spec = spec ? spec : {}
  }

  child(child: Child) {
    return this.cloneWithoutDefaultIntentHandler({child})
  }

  clone(withSpec?: PartialDocumentList): DocumentTypeListBuilder {
    const parent = super.clone(withSpec)
    const builder = new DocumentTypeListBuilder()
    builder.spec = {...this.spec, ...parent.getSpec(), ...(withSpec || {})}
    return builder
  }

  cloneWithoutDefaultIntentHandler(withSpec?: PartialDocumentList): DocumentTypeListBuilder {
    const parent = super.clone(withSpec)
    const builder = new DocumentTypeListBuilder()
    const canHandleIntent = this.spec.canHandleIntent
    const shouldOverride = canHandleIntent && canHandleIntent.identity === DEFAULT_INTENT_HANDLER
    const override = shouldOverride ? {canHandleIntent: undefined} : {}
    builder.spec = {
      ...parent.getSpec(),
      ...this.spec,
      ...(withSpec || {}),
      ...override
    }
    return builder
  }
}
