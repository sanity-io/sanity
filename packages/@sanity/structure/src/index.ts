import {uniqueId} from 'lodash'
import {SchemaType} from './parts/Schema'
import {ListBuilder, ListInput} from './List'
import {
  getDocumentTypeListItems,
  getDocumentTypeListItem,
  getDocumentTypeList,
} from './documentTypeListItems'
import {
  MenuItemBuilder,
  MenuItem,
  getOrderingMenuItemsForSchemaType,
  getOrderingMenuItem,
} from './MenuItem'
import {ListItemBuilder, ListItemInput} from './ListItem'
import {MenuItemGroup, MenuItemGroupBuilder} from './MenuItemGroup'
import {DocumentListBuilder, DocumentListInput} from './DocumentList'
import {Divider} from './StructureNodes'
import {SerializeError} from './SerializeError'
import {
  DocumentBuilder,
  PartialDocumentNode,
  documentFromEditor,
  documentFromEditorWithInitialValue,
  getDefaultDocumentNode,
} from './Document'
import {ComponentInput, ComponentBuilder} from './Component'
import {DocumentListItemBuilder, DocumentListItemInput} from './DocumentListItem'
import {Ordering} from './Sort'
import * as views from './views'
import {
  InitialValueTemplateItemBuilder,
  defaultInitialValueTemplateItems,
  menuItemsFromInitialValueTemplateItems,
} from './InitialValueTemplateItem'

const StructureBuilder = {
  defaults: getDefaultStructure,
  documentTypeList: getDocumentTypeList,
  documentTypeListItem: getDocumentTypeListItem,
  documentTypeListItems: getDocumentTypeListItems,
  document: (spec?: PartialDocumentNode) => new DocumentBuilder(spec),
  documentWithInitialValueTemplate: documentFromEditorWithInitialValue,
  defaultDocument: getDefaultDocumentNode,

  list: (spec?: ListInput) => new ListBuilder(spec),
  listItem: (spec?: ListItemInput) => new ListItemBuilder(spec),

  menuItem: (spec?: MenuItem) => new MenuItemBuilder(spec),
  menuItemGroup: (spec?: MenuItemGroup) => new MenuItemGroupBuilder(spec),
  menuItemsFromInitialValueTemplateItems,

  documentList: (spec?: DocumentListInput) => new DocumentListBuilder(spec),
  documentListItem: (spec?: DocumentListItemInput) => new DocumentListItemBuilder(spec),

  orderingMenuItem: (ordering: Ordering) => getOrderingMenuItem(ordering),
  orderingMenuItemsForType: (type: string) => getOrderingMenuItemsForSchemaType(type),

  editor: documentFromEditor,
  editorWithInitialValueTemplate: documentFromEditorWithInitialValue,

  defaultInitialValueTemplateItems,
  initialValueTemplateItem: (
    templateId: string,
    parameters?: {[key: string]: any}
  ): InitialValueTemplateItemBuilder =>
    new InitialValueTemplateItemBuilder({
      id: templateId,
      parameters,
      templateId,
    }),

  component: (spec?: ComponentInput | Function) => {
    return typeof spec === 'function'
      ? new ComponentBuilder().component(spec)
      : new ComponentBuilder(spec)
  },

  divider: (): Divider => ({id: uniqueId('__divider__'), type: 'divider'}),

  view: views,
}

function hasIcon(schemaType?: SchemaType | string): boolean {
  if (!schemaType || typeof schemaType === 'string') {
    return false
  }

  return Boolean(schemaType.icon)
}

function getDefaultStructure(): ListBuilder {
  const items = getDocumentTypeListItems()
  return new ListBuilder()
    .id('__root__')
    .title('Content')
    .items(items)
    .showIcons(items.some((item) => hasIcon(item.getSchemaType())))
}

export type {InitialValueTemplateItem} from './InitialValueTemplateItem'

export {StructureBuilder, SerializeError}
