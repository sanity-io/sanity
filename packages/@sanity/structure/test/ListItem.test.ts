import {StructureBuilder as S} from '../src'

test('builds list items with only ID and title', () => {
  expect(S.listItem({id: 'foo', title: 'Foo'}).serialize()).toMatchSnapshot()
})

test('throws if no id is set', () => {
  expect(() => S.listItem().serialize()).toThrowErrorMatchingSnapshot()
})

test('infers ID from title if not specified', () => {
  expect(S.listItem().title('Hei der').getId()).toEqual('heiDer')
  expect(S.listItem().id('zing').title('Hei der').getId()).toEqual('zing')
  expect(S.listItem().title('Hei der').id('blah').getId()).toEqual('blah')
})

test('throws if no title is set', () => {
  expect(() => S.listItem().id('foo').serialize()).toThrowErrorMatchingSnapshot()
})

test('builds list items with ID and title through setters', () => {
  expect(
    S.listItem({id: 'books', title: 'Books'}).id('authors').title('Authors').serialize()
  ).toMatchSnapshot()
})

test('builds list items with specific child (editor node)', () => {
  expect(
    S.listItem({id: 'wow', title: 'The Winds of Winter'})
      .child({id: 'foo', type: 'editor', options: {id: 'wow', type: 'book'}})
      .serialize()
  ).toMatchSnapshot()
})

test('builds list items with specific child resolver', () => {
  expect(
    S.listItem({id: 'wow', title: 'The Winds of Winter'})
      .child(() =>
        S.editor({id: 'editor', title: 'Editor', options: {id: 'docId', type: 'book'}}).serialize()
      )
      .serialize()
  ).toMatchSnapshot()
})

test('builds list items with child defined through builder', () => {
  expect(
    S.listItem({id: 'asoiaf', title: 'A Song of Ice and Fire'})
      .child(S.documentList().id('asoiaf-books').filter('author == "grrm"'))
      .serialize()
  ).toMatchSnapshot()
})

test('builds list items with specified schema type', () => {
  expect(
    S.listItem({id: 'foo', title: 'Foo'})
      .schemaType({name: 'book', type: {name: 'object'}})
      .serialize()
  ).toMatchSnapshot()
})

test('builds list items with display options (dont show icon)', () => {
  expect(S.listItem({id: 'foo', title: 'Foo'}).showIcon(false).serialize()).toMatchSnapshot()
})

test('builds list items with display options (show icon)', () => {
  expect(S.listItem({id: 'foo', title: 'Foo'}).showIcon(true).serialize()).toMatchSnapshot()
})

test('builder is immutable', () => {
  const original = S.listItem()
  expect(original.id('foo')).not.toBe(original)
  expect(original.title('foo')).not.toBe(original)
  expect(original.child(() => undefined)).not.toBe(original)
  expect(original.schemaType('foo')).not.toBe(original)
  expect(original.showIcon(true)).not.toBe(original)
})

test('getters work', () => {
  const original = S.listItem()
  expect(original.id('foo').getId()).toEqual('foo')
  expect(original.title('bar').getTitle()).toEqual('bar')
  expect(original.schemaType('baz').getSchemaType()).toEqual('baz')
  expect(original.showIcon(true).getShowIcon()).toEqual(true)
})
