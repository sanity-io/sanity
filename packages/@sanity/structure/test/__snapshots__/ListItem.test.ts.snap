// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds list items with ID and title through setters 1`] = `
Object {
  "child": undefined,
  "id": "authors",
  "schemaType": undefined,
  "title": "Authors",
  "type": "listItem",
}
`;

exports[`builds list items with child defined through builder 1`] = `
Object {
  "child": Object {
    "canHandleIntent": [Function],
    "child": [Function],
    "defaultLayout": undefined,
    "displayOptions": undefined,
    "id": "asoiaf-books",
    "initialValueTemplates": Array [],
    "menuItemGroups": Array [],
    "menuItems": Array [],
    "options": Object {
      "apiVersion": "1",
      "filter": "author == \\"grrm\\"",
    },
    "schemaTypeName": undefined,
    "title": undefined,
    "type": "documentList",
  },
  "id": "asoiaf",
  "schemaType": undefined,
  "title": "A Song of Ice and Fire",
  "type": "listItem",
}
`;

exports[`builds list items with display options (dont show icon) 1`] = `
Object {
  "child": undefined,
  "displayOptions": Object {
    "showIcon": false,
  },
  "id": "foo",
  "schemaType": undefined,
  "title": "Foo",
  "type": "listItem",
}
`;

exports[`builds list items with display options (show icon) 1`] = `
Object {
  "child": undefined,
  "displayOptions": Object {
    "showIcon": true,
  },
  "id": "foo",
  "schemaType": undefined,
  "title": "Foo",
  "type": "listItem",
}
`;

exports[`builds list items with only ID and title 1`] = `
Object {
  "child": undefined,
  "id": "foo",
  "schemaType": undefined,
  "title": "Foo",
  "type": "listItem",
}
`;

exports[`builds list items with specific child (editor node) 1`] = `
Object {
  "child": Object {
    "id": "foo",
    "options": Object {
      "id": "wow",
      "type": "book",
    },
    "type": "editor",
  },
  "id": "wow",
  "schemaType": undefined,
  "title": "The Winds of Winter",
  "type": "listItem",
}
`;

exports[`builds list items with specific child resolver 1`] = `
Object {
  "child": [Function],
  "id": "wow",
  "schemaType": undefined,
  "title": "The Winds of Winter",
  "type": "listItem",
}
`;

exports[`builds list items with specified schema type 1`] = `
Object {
  "child": undefined,
  "id": "foo",
  "schemaType": Object {
    "name": "book",
    "type": Object {
      "name": "object",
    },
  },
  "title": "Foo",
  "type": "listItem",
}
`;

exports[`throws if no id is set 1`] = `"\`id\` is required for list items"`;

exports[`throws if no title is set 1`] = `"\`title\` is required for list items"`;
